src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+using System.Globalization;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal partial class CborReader
+    {
+        public CborTag ReadTag()
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+            return tag;
+        }
+
+        public void ReadTag(CborTag expectedTag)
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            if (expectedTag != tag)
+            {
+                throw new InvalidOperationException("CBOR tag mismatch.");
+            }
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+        }
+
+        public CborTag PeekTag() => PeekTagCore(out int _);
+
+        private CborTag PeekTagCore(out int additionalBytes)
+        {
+            CborInitialByte header = PeekInitialByte(expectedType: CborMajorType.Tag);
+            return (CborTag)ReadUnsignedInteger(_buffer.Span, header, out additionalBytes);
+        }
+
+        // Additional tagged type support
+
+        public DateTimeOffset ReadDateTimeOffset()
```

Comment:
This method will recognize both unix time and string datetime formats. I'm wondering whether it might be appropriate that those be exposed as separate methods.

Reply:
Perhaps it just needs a defaulted `CborTag? expectedTag` parameter, that currently supports `null` (don't care, just do it) and the two formats.  Other inputs would be an ArgumentOutOfRangeException until/unless new DateTimeOffset-compatibile tag types get added.

Reply:
Hmm, I'm not sure I'm a big fan of that, it also means it introduces a new failure mode for tags that don't match either of the two expected values. I'd probably either leave as-is or break up into two separate methods:

```csharp
DateTimeOffset ReadDateTimeString(); // for lack of a better name for RFC3339 date strings
DateTimeOffset ReadUnixDateTime();
```
As you noted in a different comment, it should be trivial for consumers to recompose into one extension method that handles both cases.